tests/backends/storage/test_s3.py

```diff
@@ -0,0 +1,378 @@
+"""Tests for Ralph S3 storage backend"""
+
+import datetime
+import json
+import logging
+import sys
+from io import BytesIO
+from pathlib import Path
+
+import boto3
+import botocore
+import pytest
+from moto import mock_s3
+
+from ralph.conf import settings
+from ralph.exceptions import BackendException, BackendParameterException
+
+
+@mock_s3
+def test_backends_storage_s3_storage_instantiation_should_raise_exception(
+    s3, caplog
+):  # pylint:disable=invalid-name
+    """Checks that S3Storage raises BackendParameterException on failure."""
+    # Regions outside of us-east-1 require the appropriate LocationConstraint
+    s3_client = boto3.client("s3", region_name="us-east-1")
+    bucket_name = "my-test-bucket"
+    s3_client.create_bucket(Bucket=bucket_name)
+
+    error = "Not Found"
+    caplog.set_level(logging.ERROR)
+
+    with pytest.raises(BackendParameterException):  # , match=error):
+        s3()
+    logger_name = "ralph.backends.storage.s3"
+    msg = f"Unable to connect to the requested bucket: {error}"
+    assert caplog.record_tuples == [(logger_name, logging.ERROR, msg)]
+
+
+@mock_s3
+def test_backends_storage_s3_storage_instantiation_failure_should_not_raise_exception(
+    s3,
+):  # pylint:disable=invalid-name
+    """Checks that S3Storage doesn't raise exceptions when the connection is
+    successful.
+    """
+    # Regions outside of us-east-1 require the appropriate LocationConstraint
+    s3_client = boto3.client("s3", region_name="us-east-1")
+    bucket_name = "bucket_name"
+    s3_client.create_bucket(Bucket=bucket_name)
+
+    try:
+        s3()
+    except Exception:  # pylint:disable=broad-except
+        pytest.fail("S3Storage should not raise exception on successful connection")
+
+
+@mock_s3
+def test_backends_storage_s3_list_should_yield_archive_names(
+    s3, fs
+):  # pylint:disable=invalid-name
+    """Tests that given S3Service.list method successfully connects to the S3
+    storage, the S3Storage list method should yield the archives.
+    """
+    # Fix for incompatibility between moto and pyfakefs
+    # See https://github.com/spulec/moto/issues/1682
+    for module in [boto3, botocore]:
+        module_dir = Path(module.__file__).parent
+        fs.add_real_directory(module_dir, lazy_read=False)
```

Comment:
Have you tried to factor this in the test suite `setup()` function?

Reply:
The _fs_ pyfakefs plugin from pytest has a **function** scope. Best I can do is factoring the **code** in a function scope fixture, but the fixture still needs to be called for each test method where moto and fs are used together. Pytest has fs_module or fs_session with larger scope, but you can only choose one so it would have an impact on many tests.